pkg/daemon/ceph/cleanup/disk.go

```diff
@@ -0,0 +1,179 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cleanup
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/daemon/ceph/osd"
+	oposd "github.com/rook/rook/pkg/operator/ceph/cluster/osd"
+)
+
+const (
+	ddBS      = "1M"           // DD's block size
+	ddCount   = "10"           // DD runs over the first 10 offsets
+	ddFlags   = "direct,dsync" // DD's sync flags"
+	ddIf      = "/dev/zero"
+	ddUtility = "dd"
+)
+
+var (
+	logger = capnslog.NewPackageLogger("github.com/rook/rook", "cleanup")
+)
+
+// DiskSanitizer is simple struct to old the context to execute the commands
+type DiskSanitizer struct {
+	context     *clusterd.Context
+	clusterName string
+	clusterFSID string
+	disk        string
+}
+
+// NewDiskSanitizer is function that returns a full filled DiskSanitizer object
+func NewDiskSanitizer(context *clusterd.Context, clusterName, clusterFSID string) *DiskSanitizer {
+	return &DiskSanitizer{
+		context:     context,
+		clusterName: clusterName,
+		clusterFSID: clusterFSID,
+	}
+}
+
+// StartSanitizeDisks main entrypoint of the cleanup package
+func StartSanitizeDisks(sanitizer *DiskSanitizer) {
+	// LVM based OSDs
+	osdLVMList, err := osd.GetCephVolumeLVMOSDs(sanitizer.context, sanitizer.clusterName, sanitizer.clusterFSID, "", false, false)
+	if err != nil {
+		logger.Errorf("failed to list lvm osd(s). %v", err)
+	} else {
+		// Start the sanitizing sequence
+		sanitizer.sanitizeLVMDisk(osdLVMList)
+	}
+
+	// Raw based OSDs
+	osdRawList, err := osd.GetCephVolumeRawOSDs(sanitizer.context, sanitizer.clusterName, sanitizer.clusterFSID, "", "", false)
+	if err != nil {
+		logger.Errorf("failed to list raw osd(s). %v", err)
+	} else {
+		// Start the sanitizing sequence
+		sanitizer.sanitizeRawDisk(osdRawList)
+	}
+}
+
+func (s *DiskSanitizer) sanitizeRawDisk(osdRawList []oposd.OSDInfo) {
+	// Initialize work group to wait for completion of all the go routine
+	var wg sync.WaitGroup
+
+	for _, osd := range osdRawList {
+		// Assign disk to sanitize
+		s.disk = osd.BlockPath
+
+		logger.Infof("sanitizing osd %d disk %q", osd.ID, osd.BlockPath)
+
+		// Increment the wait group counter
+		wg.Add(1)
+
+		// Put each sanitize in a go routine to speed things up
+		go s.executeSanitizeCommand(s.disk, &wg)
+	}
+
+	wg.Wait()
+}
+
+func (s *DiskSanitizer) sanitizeLVMDisk(osdLVMList []oposd.OSDInfo) {
+	// Initialize work group to wait for completion of all the go routine
+	var wg sync.WaitGroup
+	pvs := []string{}
+
+	for _, osd := range osdLVMList {
+		// Increment the wait group counter
+		wg.Add(1)
+
+		// Assign disk
+		s.disk = osd.BlockPath
```

Comment:
The disk needs to be passed as a param to `wipeLVM()`, otherwise all the goroutines will overwrite `s.disk` and end up working on the same device. Or maybe we shouldn't worry about the complexity with the goroutines for now.

Reply:
This is just used for `returnPVDevice` and `wipeLVM()` only needs the osd ID so no overwrite will happen.
It hasn't happen actually.

Reply:
Then it can just be a param to `returnPVDevice(), right`? But I do see now why it's not an issue in the goroutine if it's only used by returnPVDevice.